### ISSUE TITLE
Resuming from checkpoint doubles time to train each step?

### ISSUE THREAD

**Issue by sdman88** [tdrussell/diffusion-pipe#127]
Hi, I noticed when using the --resume from checkpoint tag it seems to significantly increase the time to train additional steps by about double the normal amount of time when not resuming training from checkpoint.  Nothing has changed between my settings or dataset either.  I'm running on 12GB vram 3060 if that matters, image only training on 35 512x512 images. When I first started the training from scratch it was about 3-4 mins per step and after resuming it was about 7 - 8 mins a step. Any advice? Is this normal or unexpected behavior? Thanks

**Comment by tdrussell** [comment]
Just to clarify, you mean several minutes per single training step, not epoch? And I assume you are using WSL? Because HunyuanVideo is 13b parameters, at fp8 the model weights alone are 13GB. So the windows Nvidia driver is doing its memory offloading thing, which is the only reason you're able to run it at all. I would not be surprised if the slowdown caused by the driver offloading feature is unpredictable.

**Comment by sdman88** [comment]
Right per training step not per epoch but it also is affecting epochs since it takes twice as long to train overall just about when resuming from checkpoint vs from scratch for me. But yes using WSL on windows 11 machine with cuda installed so the offloading I get since too large to fit in 12gb vram alone, but I trained 10 epochs nonstop and the time per step count remained pretty much constant but as soon as I stopped it and then reloaded from checkpoint to train more epochs, it immediately about doubled the time per step and it remained constant per each step and epoch compared to from scratch and nothing changed between my dataset or settings like I mentioned, does reloading from checkpoint put more load on vram/cpu compared to from scratch or something?

**Comment by tdrussell** [comment]
I regularly use the resume_from_checkpoint feature and I have never seen any slowdown. Speed and memory use upon resuming are always exactly the same. But I am on linux, so there is no automatic VRAM offloading in the driver.

Have you tried reproducing the slowdown? There is a checkpoint_every_n_minutes setting, you can cause it to checkpoint training immediately and then resume. And what if you train a smaller model that fits in your VRAM? I suspect that any slowdown is related to the VRAM offloading, but I doubt there's anything I can do about that. The exact memory access patterns could affect when and how the driver is moving data to/from VRAM, and like I said I would not be surprised if this is nondeterministic.

FYI I want to eventually try implementing block swapping, this would give a principled and configurable way to offload layers to RAM that should be better than relying on a GPU driver feature to do it automatically.

**Comment by sdman88** [comment]
Hmm, interesting you're not experiencing the slowdown at all, yeah I've been able to reproduce the slowdown every time I try training, if I stop to test the epochs that have been trained already and then resume training afterwards it always results in training taking almost double amount of time as before. I'm wondering if it's something that got messed up during installation of WSL and the diffusion-pipe folder, I noticed it put the diffusion-pipe folder on my windows C drive which I don't know if that's normal since I don't have prior WSL experience. Wondering if it's loading something into memory from the C drive folder that causes additional load on VRAM thus increasing the amount of time taken?  Cause if it was an offloading issue I would assume I would see variations in time to train each step while training without stopping but it always remains consistent up until I resume my training then it always doubles amount of time taken and remains constant without much fluctuation.  Should I be using the --regenerate_cache tag when resuming from checkpoint or is that only necessary if changes are made I assume? Thanks for that help and the info, I'm not exactly sure what all the blocking stuff means though I'm not super knowledgeable on that aspect

**Comment by nikolaiusa** [comment]
did you manage to figure it out? I have the same problem with speed when resuming training hidream. 7sec/step vs 45sec/step 4090 23.5gb mem use

**Comment by nikolaiusa** [comment]
after 
[2025-08-24 11:54:25,330] [INFO] [logging.py:107:log_dist] [Rank 0] step=17402, skipped=0, lr=[5e-05], mom=[0.0]
steps: 17402 loss: 0.2046 iter time (s): 102.755 samples/sec: 0.039

before: +-7samples/sec
